### PR TITLE
Target multi-entity sub-sensors when adding automations

### LINKED ISSUE #1263: [Bug] No way to specify sensor to attach aht20 component automations
### What happened?

There isn't a "Add Automation" option for the individual Temperature and Humidity parts of the config, instead that option exist only for the component. 

1. In Device Navigator, select "AHT10 Temperature+Humidity Sensor"
2. Scroll down to Automations and click "Add Automation"
3. Set up the trigger:

    What should this automation react to? → A configured component
    Which configured component? → aht20 #There is no Temperature or Humidity specific option
    Which trigger? → Sensor → On Value Range

Image

 When setting up an automation that way, the flow offers no way to specify which sensor it belongs to and produces an invalid YAML with the follow error:

`[on_value_range] is an invalid option for [sensor.aht10]. Please check the indentation.`

```yaml
...
sensor:
  - platform: aht10
    i2c_id: i2c_1
    variant: AHT20
    humidity:
      id: aht20_humidity
      name: Kit Humidity
      device_class: humidity
      unit_of_measurement: "%"
      accuracy_decimals: 2
      force_update: false
      state_class: measurement
    temperature:
      id: aht20_temperature
      name: Kit Temperature
      device_class: temperature
      unit_of_measurement: °C
      accuracy_decimals: 2
      force_update: false
      state_class: measurement
    id: aht20
    on_value_range:
      - above: 10
        below: 20
        then:
          - light.toggle: onboard_rgb_led
...
```

### Dashboard version

2026.5.3

### How did you install it?

Home Assistant App (preview toggle)

### Where is it running?

Home Assistant App

### Relevant logs / errors

```shell

```

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by bdraco**
# What does this implement/fix?

Backend half of the AHT10-style automation bug. A multi-entity platform component (e.g. `sensor: - platform: aht10`) exposes its readings as nested `platform_type` sub-blocks (`temperature:` / `humidity:`), each its own entity with an `id`. Entity triggers like `on_value_range` belong on those sub-blocks, not the platform item; the old flow only offered the container instance and spliced `on_value_range:` directly under the platform item, producing invalid YAML with no way to pick which reading.

This makes `automations/get_available` surface each ided sub-entity as its own bare-domain instance (`component_id: "sensor"`, `parent_id` set to the container) and marks the platform item `is_entity_container: true`. The writer resolves a sub-entity id to its parent + sub-key and splices the handler under the nested block (sibling of the nested `id:`); the parser reads such handlers back to the same sub-entity location, so add / delete round-trip. A single-reading sensor (`sensor.adc`) is unchanged.

Sequenced behind the writer split that already merged (#1265).

**Related issue or feature (if applicable):**

- fixes https://github.com/esphome/device-builder/issues/1263

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue) — `bugfix`
- [ ] New feature (non-breaking change which adds functionality) — `new-feature`
- [ ] Enhancement to an existing feature — `enhancement`
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected) — `breaking-change`
- [ ] Refactor (no behaviour change) — `refactor`
- [ ] Documentation only — `docs`
- [ ] Maintenance / chore — `maintenance`
- [ ] CI / workflow change — `ci`
- [ ] Dependencies bump — `dependencies`

## Frontend coordination

- [ ] No frontend change needed
- [ ] Companion frontend PR: esphome/device-builder-frontend#<number>

`AvailableComponentInstance` gains two optional fields (`is_entity_container`, `parent_id`); existing instances keep their wire shape via defaults. A companion frontend PR (grouped sub-entity picker that hides the container and offers entity triggers on the sub-entities) is forthcoming; the new fields are optional so the frontend degrades gracefully until then.

## Checklist

- [x] The code change is tested and works locally.
- [x] Pre-commit hooks pass (`ruff`, `codespell`, yaml/json/python checks).
- [x] Tests have been added or updated under `tests/` where applicable.
- [x] `components.index.json` / `definitions/components/*.json` have **not** been hand-edited (regenerate via `script/sync_components.py` if a sync is needed).
- [x] Architecture-level changes are reflected in `docs/ARCHITECTURE.md` and/or `docs/API.md`.

### CODE REVIEW

**Review by esphbot — CHANGES_REQUESTED** [review]
Blocking issues found — see the review comment above.

File: esphome_device_builder/controllers/automations/writing.py [hunk 1]
```diff
@@ -154,11 +162,11 @@ def _upsert_component_on(
...
     """Splice an inline ``on_*:`` handler under a configured component."""
+    target = resolve_component_target(yaml_text, location.component_id)
+    if target is not None and target.is_sub_entity:
+        return _upsert_subentity_on(yaml_text, tree, location, target)
     instance_domain = _component_domain_from_yaml(yaml_text, location)
-    trigger = catalog.trigger_by_id(f"{instance_domain}.{location.trigger}")
-    if trigger is None:
-        msg = f"Unknown trigger id {location.trigger!r} on component {location.component_id!r}"
-        raise CommandError(ErrorCode.INVALID_ARGS, msg)
+    trigger = _require_trigger(instance_domain, location)
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome_device_builder/controllers/automations/writing.py:169]: Fixed. `_upsert_component_on` now reuses the already-resolved target: `instance_domain = target.domain if target is not None else _component_domain(location)`, so the YAML is parsed once. Removed the now-dead `_component_domain_from_yaml`.

File: esphome_device_builder/controllers/automations/writing.py [hunk 2]
```diff
@@ -505,6 +541,9 @@ def _delete_component_on(
     location: ComponentOnLocation,
 ) -> tuple[str, YamlDiff]:
     """Drop an inline ``on_*:`` handler from a configured component."""
+    target = resolve_component_target(yaml_text, location.component_id)
+    if target is not None and target.is_sub_entity:
+        return _delete_subentity_on(yaml_text, location, target)
     instance_domain = _component_domain_from_yaml(yaml_text, location)
     if location.index is not None:
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome_device_builder/controllers/automations/writing.py:548]: Fixed the same way in `_delete_component_on` (reuses `target.domain`, no second parse).

File: esphome_device_builder/controllers/automations/writing.py [hunk 3]
```diff
@@ -608,6 +671,23 @@ def _delete_api_action(
...
+        raise CommandError(ErrorCode.INVALID_ARGS, msg)
+    return trigger
+
+
+def _subentity_context(target: ComponentTarget) -> tuple[str, str, str]:
+    """Unpack a sub-entity target's parent context (guaranteed by ``is_sub_entity``)."""
+    assert target.parent_domain is not None
+    assert target.parent_id is not None
+    assert target.sub_key is not None
+    return target.parent_domain, target.parent_id, target.sub_key
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome_device_builder/controllers/automations/writing.py]: Fixed. `_subentity_context` now raises `CommandError(INVALID_ARGS)` instead of asserting, so a broken invariant still surfaces under `python -O` rather than leaking None into the splice. Added a test for the incomplete-target path.

**Comment by bdraco** [comment]
Thanks — addressed the blocking + suggestion items in the latest push (now rebased on main):

- **🟡 is_entity_container capability vs presence**: now presence-based. `is_container = bool(subs)` where `subs` are the ided sub-entities actually surfaced, so an id-less multi-entity platform keeps its plain instance instead of being hidden with nothing to target. Added a test (`test_get_available_idless_multi_entity_is_not_a_container`).
- **🟢 resolve_component_target duplicated the walk**: extracted `_iter_instance_targets` as the single document walk; `resolve_component_target` and `_iter_component_instances` are now thin consumers, so the parse and writer walks can't drift.
- **🟢 not-found splice branches untested**: added tests for the `res is None` raise paths in `_upsert_subentity_on` / `_delete_subentity_on` (missing parent and missing sub-block) plus the traversal-id guard; changed-line patch coverage is back to 100%.
- **Silent-failure (assert)**: `_subentity_context` now raises `CommandError` instead of asserting, so the invariant holds under `python -O`; covered by a test.

Also resolved Copilot's double-parse note by reusing the resolved target's domain and removing `_component_domain_from_yaml`.

**Review by esphbot — APPROVED** [review]
No blocking issues found.

**Review by esphbot — APPROVED** [review]
No blocking issues found.

**Comment by bdraco** [comment]
Thanks — both 🟢 cleanups done:

1. **Shared walk widening the action-field pass**: `_parse_component_action_fields` now iterates `_iter_instance_targets` and skips `is_sub_entity`, so its scope is explicitly top-level-only regardless of what the shared walk yields (no longer relying on the bare-domain body being absent). The inline-`on_*` pass keeps the sub-entities it needs for round-trip.
2. **Redundant platform_type lookup**: collapsed `_is_subentity_block` to an early non-nested return plus a single bound `platform_type` lookup.

Coverage stays at 100% on the changed lines.

**Review by esphbot — APPROVED** [review]
No blocking issues found.
